### PR TITLE
Add initial Cypress tests for pivot table (ref: #14049)

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
READY

### What does this PR accomplish?
- Adds an initial batch of Cypress tests for pivot table functionality as described in #14049

### How to test this?
- All tests should pass in CI, because no test is skipped.

### Additional notes:
- Sidebar opening doesn't have any XHR associated with it that we can wait for. I tried to rely on UI, but I'm concerned that different visualization icons and/or drag handles may not be ready and the tests may flake.
- Cypress occasionally failed for me locally saying that icon/drag handle is not visible.
- If that presents a problem, I can remove the `.should("be.visible)` and add `{force: true}` to an associated action but that is an anti-pattern.
- FWIW, so far every CI run was successful.

### Stress-testing in GitHub CI:
- I was using OSS uberjar generated from 6697758c409d5ec030ad3da456910d49d5381525
- [20/20 runs](https://github.com/nemanjaglumac/metabase-tests/actions/runs/416375653) were successful ✅

### CODE REVIEW

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:55]: @paulrosenzweig please let me know if this is too specific. I could just test that these options are present, without asserting on their order.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:39]: It'd be nice if we could skip this UI configuration and instead push the question we want in the URL. IIRC Tom tried that once and encountered issues loading the needed frontend code in Cypress. Maybe we could write a simple helper that takes the JSON question we want and base64 encodes it for the URL.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:44]: Why do we need to wait for this rather than looking for the UI?

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:54]: There's probably no reason to assert the order, but I don't think it hurts.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:73]: Looks good!

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:39]: (That's unrelated to this PR)

**Review by paulrosenzweig — APPROVED** [review]

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:44]: Just as a precaution if it takes longer for calculation to finish in CI. Relying on UI leads to all sorts of issues and race conditions.
This simple test could probably go without waiting for anything, though.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:39]: That would be awesome, but there's an alternative for this case.
I can use API to:
1. store this question (with `bar` visualization),
2. Open it and use UI to change the visualization to pivot table (that would still make it "ad-hoc", right?)
3. Continue asserting on UI

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:73]: Thank you.
This is just the first test. I'll use this PR to group them all and then merge them to `master`.

@paulrosenzweig 
Will it make your life easier if I separate all tests in separate PRs (that will use this one as a base)?

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: > that would still make it "ad-hoc", right?

Unfortunately, it wouldn't since the query is unchanged. You could add another breakout or change the aggregate to switch to ad-hoc, but that point it's probably simpler to just use the UI the whole way.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: No more waiting :)

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: 👍 one PR works for me

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: See [this line](https://github.com/metabase/metabase/pull/14051/files#diff-d5b5ec136d796378e5808beb6b14135815b5dbcc1fd7b47a06c61b7ef042d5a7R12)

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:44]: This is needed until I merge Cypress metadata extraction updates from the `release-x.37.x` branch.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:90]: Note to myself - Fix case on `pivot` to `Pivot`

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: [Done](https://github.com/metabase/metabase/commit/6697758c409d5ec030ad3da456910d49d5381525). Rebased and refactored this function.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:152]: @paulrosenzweig this function works fine for an initial state when we want to drag one field "on top of each other" (it gives us fail-proof pointer/location). However, if you think that we should add more drag and drop scenarios, I'd probably have to either expand it or think of some other ways to accomplish that.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:74]: @paulrosenzweig 

Continuing from [this comment](https://github.com/metabase/metabase/pull/14051/files#r541291213) - are there other scenarios we should test?
- Drag row field into columns? (that gave me empty field values for the whole table)
- Assert that the `count` cannot be dragged anywhere?
- ... add your own scenario

Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:74]: I don't think we need to. This checks that the setting interaction works, we requery for new data, and we update the chart. Also, the UI here might change a bit still so it'd better not too have too many tests that will break.

> that gave me empty field values for the whole table
Whoops! I either need to prevent that move or fix it's display.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (paulrosenzweig) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js]: I wasn't expected that to run as an ad-hoc query, but I see that it does!

I think that behavior could change in the future. We don't need to run that as ad-hoc since the display isn't part of the query. If we refine that logic to prioritize running saved cards, this test will no longer use an ad-hoc query.

File: frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/visualizations/pivot_tables.cy.spec.js:74]: Ok, so this can be merged and please let me know if you think of any more test cases, or need help reproing anything related.

Thank you for reviewing this.

**Pull request merged**
